### PR TITLE
PostHog demo environment

### PR DESCRIPTION
**Pull request by Twixes**
## Changes

This adds a new mode of operation to the app: PostHog Demo™.
Enabled with env var `DEMO`, it merges the normal signup and login flows into a "demo signin", i.e. you create an account without a password (when you first enter the demo environment with that email) and log in also without a password (if entering with an already used email). On account creation, you're plopped into a fresh demo project. The announcement bar at the top shows a special demo mode message. As for analytics, demo mode is a new `realm` (to track it while distinguishing it from production instances): `demo`.

Demo of the demo:
demo demo

## How did you test this code?

Spun up a PostHog Demo™ instance with the hobby script, available at https://demo.posthog.com (just as a proof of concept, will need to move away from my personal DigitalOcean account later on).

### CODE REVIEW

File: posthog/api/signup.py [hunk 1]
```diff
@@ -44,16 +46,32 @@ def validate_password(self, value):
...
+                **validated_data,
+                is_staff=is_instance_first_user,
+            )
+        else:
+            # If there's an email collision in signup in the demo environment, we treat it as a login
+            is_instance_first_user = False
+
+            self._user = already_existing_demo_user
+            self._organization = already_existing_demo_user.organization
+            self._team = already_existing_demo_user.team
```
Comment (mariusandra) [inline posthog/api/signup.py:73]: A nit perhaps, but could this be refactored to:

```
if settings.DEMO:
    # override everything
else:
    # normal flow
```

It's kind of hard to follow along, and such critical code as the signup flow should be made as mistake-proof as possible. Maybe already just flipping the if `if not already_existing_demo_user:` to `if already_existing_demo_user:` will make this cleaner?

File: posthog/utils.py [hunk 2]
```diff
@@ -601,10 +601,12 @@ def is_clickhouse_enabled() -> bool:
 
 def get_instance_realm() -> str:
     """
-    Returns the realm for the current instance. `cloud` or `hosted` or `hosted-clickhouse`.
+    Returns the realm for the current instance. `cloud` or 'demo' or `hosted` or `hosted-clickhouse`.
     """
     if settings.MULTI_TENANCY:
         return "cloud"
+    if settings.DEMO:
```
Comment (mariusandra) [inline posthog/utils.py:608]: Nittiest of nits, but `if` -> `elif` to keep tactical formation.

**Review by mariusandra — APPROVED** [review]
Found some things to nit about, but looks good.

The only question I have is if we actually want to enable this? Probably, but then we must make sure it's always up to date with some playable data. Enabling the "make some noise" plugin might be enough even?

**Comment by jamesefhawkins** [comment]
I've left [some comments](https://github.com/PostHog/posthog.com/pull/2661#issuecomment-1000198985) on why I think we can't make this somewhere users go yet

**Comment by guidoiaquinti** [comment]
Overall LGTM but I'll appreciate also another pair of 👀 from someone of my team. Also, what's the long term plan for this? Do we have a plan to make sure:

* the instance is up-to-date
* we have proper monitoring/alerting setup
* etc..

Thank you!

File: frontend/src/layout/navigation/TopBar/Announcement.tsx [hunk 3]
```diff
@@ -1,20 +1,30 @@
...
 
-    let message = <Fragment />
-    if (shownAnnouncementType === AnnouncementType.CloudFlag && cloudAnnouncement) {
+    let message: JSX.Element | undefined
+    if (preflight?.demo) {
+        message = (
+            <b>
+                Welcome to PostHog's demo environment. To level up,{' '}
+                <a href="https://posthog.com/signup/self-host">easily deploy PostHog yourself</a> or{' '}
+                <a href="https://app.posthog.com/signup">sign up for a free PostHog Cloud account</a>.
```
Comment (tiina303) [inline frontend/src/layout/navigation/TopBar/Announcement.tsx:24]: Nit: https://posthog.com/signup covers both self-hosting and cloud and we should likely just redirect them there, and that page can evolve and give all the guidance. "easily deploy" might mislead folks to thinking it's easier to self-host compared to cloud, which isn't true. Related Q: do we want that page to highlight the option to check out the demo too?

**Review by tiina303 — APPROVED** [review]
lgtm

File: posthog/api/signup.py (hunk 1 above)
Comment (Twixes) [reply posthog/api/signup.py]: Agreed, refactored the demo handling to method `enter_demo()`.

File: frontend/src/layout/navigation/TopBar/Announcement.tsx (hunk 3 above)
Comment (Twixes) [reply frontend/src/layout/navigation/TopBar/Announcement.tsx]: Good point, simplified to a single link to that "Get started" page with copy "deploy your own PostHog instance or sign up for PostHog Cloud". I think if we go forward with the demo env, it should be somewhere on the website, but the homepage seems like the clearest place to expose it, so I'll stay with that for now (https://github.com/PostHog/posthog.com/pull/2661).

**Comment by Twixes** [comment]
Good questions @guidoiaquinti I don't have a great story for deployment _yet_ (the current MVP at demo.posthog.com rans with the hobby script), but it can be just a self-hosted PostHog instance (probably AWS or DO) pinned to `posthog/posthog:latest` – the only difference from a production instance being env var `DEMO=1` (talk about dogfooding!). Should be pretty simple to make the instance redeploy when `latest` is updated, or maybe even just once a day to keep things light. I don't think alerting is a blocker here, but should we ever want it, the instrumentation is the same as Cloud.

**Comment by tiina303** [comment]
> Should be pretty simple to make the instance redeploy when latest is updated, or maybe even just once a day to keep things light.

yes it's simple for the deployment side. I'd suggest following latest the same as cloud (which likely means you'll want to have the workflow in this repo). I'd pick DO since it's the easiest to manage, the values file should likely live in the vpc repo (so we can define stuff we don't want public, e.g sentry, google sign on). Reach out in the platform channel if we can help with anything.
